File: tools/drives/status-badge-cross-focus.py
"""The state matrix: the remove control of a status badge holding keyboard focus.

    .venv/bin/python tools/qa.py --page status-badge \
      --drive tools/drives/status-badge-cross-focus.py --shot /tmp/status-badge-cross-focus.png

The twin of `tools/drives/upstream/status-badge-cross-focus.js`. Upstream the ring is
`focus-visible:ring-2 focus-visible:ring-offset-2` of `REMOVE_CONTROL`, which a browser only
paints when it decides the focus came from a key; here rule 5 says the same thing outright, so
the badge is given focus by a key reason and the ring is painted around the cross's own box.
"""
from __future__ import annotations

from qtpy import QtCore

from sg_widgets_qt.widgets.status_badge import StatusBadge


def drive(page, wait, find, prefs) -> dict:  # noqa: ARG001
    badges = [b for b in find(StatusBadge, all=True) if b.removable and b.variant == "both"]
    if not badges:
        return {"verdict": "FAIL the page drew no removable badge"}
    badge = badges[0]
    rest = badge.grab().toImage()
    badge.setFocus(QtCore.Qt.FocusReason.TabFocusReason)
    wait(250)
    focused = badge.grab().toImage()
    ringed = focused != rest
    return {
        "verdict": (
            "PASS the cross takes the focus ring on a keyboard focus"
            if badge.keyboard_focus and ringed
            else "FAIL the cross drew no ring on a keyboard focus"
        ),
        "seen": {
            "keyboard_focus": badge.keyboard_focus,
            "repainted": ringed,
            "cross": list(badge._cross_box().getRect()),
        },
    }

File: tools/drives/upstream/status-badge-cross-focus.js
// The remove control of a status badge, holding focus, for the shot the port is read against.
//
//   pnpm qa --start --path /widgets/status-badge/ --framework react \
//     --drive .../status-badge-cross-focus.js --shot /tmp/ref/status-badge-cross-focus.png
//
// `focus-visible` follows the browser's own heuristic, so what this state shows for certain is
// where the cross sits and what it is; the ring is read against `REMOVE_CONTROL` in the port's
// own drive.
const box = $('[data-demo="removable"]');
if (!box) return { verdict: 'FAIL no removable demo on the page' };
box.scrollIntoView({ block: 'center' });
await wait(200);
const cross = $('[data-slot="status-badge-remove"]', box);
if (!cross) return { verdict: 'FAIL no cross on a removable badge' };
cross.focus();
await wait(300);
return {
  verdict:
    document.activeElement === cross
      ? 'PASS the cross holds focus'
      : 'FAIL the cross did not take focus',
  where: cross.getBoundingClientRect().toJSON(),
};
